crates/telio-batcher/src/batcher.rs

```diff
@@ -0,0 +1,404 @@
+use futures::future::BoxFuture;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::{collections::HashMap, sync::Arc};
+
+use telio_utils::telio_log_debug;
+use tokio::time::sleep_until;
+
+type Action<V, R = std::result::Result<(), ()>> =
+    Arc<dyn for<'a> Fn(&'a mut V) -> BoxFuture<'a, R> + Sync + Send>;
+
+pub struct Batcher<K, V> {
+    actions: HashMap<K, (BatchEntry, Action<V>)>,
+}
+
+struct BatchEntry {
+    deadline: tokio::time::Instant,
+    interval: std::time::Duration,
+    threshold: std::time::Duration,
+}
+
+impl<K, V> Default for Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K, V> Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    pub fn new() -> Self {
+        Self {
+            actions: HashMap::new(),
+        }
+    }
+
+    /// Batching works by sleeping until the nearest future and then trying to batch more actions
+    /// based on the threshold value. Higher delay before calling the function will increase the chances of batching
+    /// because the deadlines will _probably_ be in the past already.
+    pub async fn get_actions(&mut self) -> Option<Vec<(K, Action<V>)>> {
+        let mut batched_actions: Vec<(K, Action<V>)> = vec![];
+
+        match self.actions.len() {
+            0 => {
+                return None;
+            }
+            _ => {
+                let actions = &mut self.actions;
+
+                if let Some(closest_entry) = actions.values().min_by_key(|entry| entry.0.deadline) {
```

Comment:
NIT: maybe it makes sence to create (deadline, key) sorted vector and reuse that, you would avoid double iteration, and could also do an early return in batching step

Reply:
It's a bit more difficult since each entry has a `threshold` as well. Sure we can calculate entries as `deadline - threshold` and then sort. That _could_ work as expected but I can't prove it.

Reasoning that we hit the closest deadline and then aggregate based on threshold is very easy to reason about and to prove that it works.

However I think it's a good point to leave a comment on to optimize once we maybe have more tests regarding the subject as now basically there isn't a test that shows it works actually :D

Reply:
I left the comment for the future-me regarding your comment if you're ok with that

Reply:
It's a nit for a reason :D